File: src/main/java/gg/mineral/server/tick/TickThreadFactory.java
package gg.mineral.server.tick;

import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;

public class TickThreadFactory implements ThreadFactory {
    public static final TickThreadFactory INSTANCE = new TickThreadFactory();
    private final AtomicInteger threadCounter = new AtomicInteger();

    @Override
    public Thread newThread(Runnable runnable) {
        Thread thread = new Thread(runnable, "Mineral-tick-" + threadCounter.getAndIncrement());
        thread.setPriority(Thread.MAX_PRIORITY);
        thread.setUncaughtExceptionHandler((t, e) -> {
            System.err.println("Uncaught exception in thread " + t.getName());
            e.printStackTrace();
        });
        return thread;
    }
}

File: server/src/main/java/gg/mineral/server/tick/TickThreadFactory.kt
package gg.mineral.server.tick

import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger

class TickThreadFactory : ThreadFactory {
    private val threadCounter = AtomicInteger()

    override fun newThread(runnable: Runnable): Thread {
        val thread = Thread(runnable, "Mineral-tick-" + threadCounter.getAndIncrement())
        thread.priority = Thread.MAX_PRIORITY
        thread.uncaughtExceptionHandler =
            Thread.UncaughtExceptionHandler { t: Thread, e: Throwable ->
                System.err.println("Uncaught exception in thread " + t.name)
                e.printStackTrace()
            }
        return thread
    }

    companion object {
        val INSTANCE: TickThreadFactory = TickThreadFactory()
    }
}
